core/src/main/java/io/grpc/util/RoundRobinLoadBalancerFactory2.java

```diff
@@ -0,0 +1,303 @@
+      for (Subchannel subchannel : getSubchannels()) {
+        subchannel.shutdown();
+      }
+    }
+
+    /**
+     * Updates picker with the list of active subchannels (state == READY).
+     */
+    private void updatePicker(@Nullable Status error) {
+      List<Subchannel> activeList = filterNonFailingSubchannels(getSubchannels());
+      helper.updatePicker(new Picker(activeList, error));
+    }
+
+    /**
+     * Filters out non-ready subchannels.
+     */
+    private static List<Subchannel> filterNonFailingSubchannels(
+        Collection<Subchannel> subchannels) {
+      List<Subchannel> readySubchannels = new ArrayList<Subchannel>();
+      for (Subchannel subchannel : subchannels) {
+        if (getSubchannelStateInfoRef(subchannel).get().getState() == READY) {
+          readySubchannels.add(subchannel);
+        }
+      }
+      return readySubchannels;
+    }
+
+    /**
+     * Converts list of {@link ResolvedServerInfoGroup} to {@link EquivalentAddressGroup} set.
+     */
+    private static Set<EquivalentAddressGroup> resolvedServerInfoGroupToEquivalentAddressGroup(
+        List<ResolvedServerInfoGroup> groupList) {
+      Set<EquivalentAddressGroup> addrs = new HashSet<EquivalentAddressGroup>();
+      for (ResolvedServerInfoGroup group : groupList) {
+        for (ResolvedServerInfo server : group.getResolvedServerInfoList()) {
+          addrs.add(new EquivalentAddressGroup(server.getAddress()));
+        }
+      }
+      return addrs;
+    }
+
+    /**
+     * If all subchannels are TRANSIENT_FAILURE, return the Status associated with an arbitrary
+     * subchannel otherwise, return null.
+     */
+    @Nullable
+    private Status getAggregatedError() {
+      Status status = null;
+      for (Subchannel subchannel : getSubchannels()) {
+        ConnectivityStateInfo stateInfo = getSubchannelStateInfoRef(subchannel).get();
+        if (stateInfo.getState() != TRANSIENT_FAILURE) {
+          return null;
+        } else {
+          status = stateInfo.getStatus();
+        }
+      }
+      return status;
+    }
+
+    @VisibleForTesting
+    Collection<Subchannel> getSubchannels() {
+      return subchannels.values();
+    }
+
+    private static AtomicReference<ConnectivityStateInfo> getSubchannelStateInfoRef(
+        Subchannel subchannel) {
+      return checkNotNull(subchannel.getAttributes().get(STATE_INFO), "STATE_INFO");
+    }
+
+    private static <T> Set<T> setsDifference(Set<T> a, Set<T> b) {
+      Set<T> aCopy = new HashSet<T>(a);
+      aCopy.removeAll(b);
+      return aCopy;
+    }
+  }
+
+  @VisibleForTesting
+  static class Picker extends SubchannelPicker {
+    @GuardedBy("this")
+    final Iterator<Subchannel> subchannelIterator;
+    @Nullable
+    final Status status;
+
+    private final boolean empty;
+
+    Picker(List<Subchannel> list, @Nullable Status status) {
+      this.empty = list.isEmpty();
+      this.subchannelIterator = new CycleIterator<Subchannel>(list);
```

Comment:
Simply saving the list and keeping an index for round-robin would result in less lines of code, and allow tests to examine the list directly.